src/mame/tvgames/generalplus_gpl16250_rom.cpp

```diff
@@ -607,15 +635,109 @@ void beijuehh_game_state::machine_reset()
+	// portd is very busy on startup
 
+	m_maincpu->porta_out().set(FUNC(gameu_handheld_game_state::gameu_porta_w));
+	m_maincpu->portb_out().set(FUNC(gameu_handheld_game_state::gameu_portb_w));
+	m_maincpu->portc_out().set(FUNC(gameu_handheld_game_state::gameu_portc_w));
+	m_maincpu->portd_out().set(FUNC(gameu_handheld_game_state::gameu_portd_w));
+
+	m_screen->set_refresh_hz(30); // too fast at 60, but maybe it's for other reasons?
+	m_screen->set_visarea(0, (160)-1, 0, (128)-1); // appears to be the correct resolution for the LCD panel
+}
 
 void gormiti_game_state::machine_reset()
 {
 	gcm394_game_state::machine_reset();
 	m_maincpu->set_alt_tile_addressing_hack(1);
 }
 
+uint16_t gameu_handheld_game_state::cs0_r(offs_t offset)
+{
+	return m_romregion[(offset & 0x00fffff) + m_upperbase];
+}
+
+void gameu_handheld_game_state::gameu_porta_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	logerror("%s: porta write %04x\n", machine().describe_context(), data);
+	m_porta_data = data;
+}
+
+void gameu_handheld_game_state::gameu_portb_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	logerror("%s: portb write %04x\n", machine().describe_context(), data);
+	m_portb_data = data;
+}
+
+void gameu_handheld_game_state::gameu_portc_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	logerror("%s: portc write %04x\n", machine().describe_context(), data);
+	m_portc_data = data;
+}
+
+void gameu_handheld_game_state::gameu_portd_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	// hacky, maybe we need better direction/attribute handling on the ports in the core?
+	m_portd_data = data;
+	//int pc = m_maincpu->pc();
+	//if ((pc != 0x2b49) && (pc != 0x2b34) && (pc != 0x2b8b) && (pc != 0x2bc0))
+	{
+		logerror("%s: portd write %04x %04x\n", machine().describe_context(), data, mem_mask);
+
+		m_upperbase = (data & 0xfc00) >> 10;
+		m_upperbase *= 0x40000;
+	}
+
+}
+
+void gameu_handheld_game_state::machine_reset()
+{
+	gcm394_game_state::machine_reset();
+	m_maincpu->set_alt_tile_addressing_hack(1);
+	m_upperbase = 0;
+}
+
+void gameu_handheld_game_state::init_gameu()
+{
+	uint16_t *ROM = (uint16_t*)memregion("maincpu")->base();
+	int size = memregion("maincpu")->bytes();
+
+	for (int i = 0; i < size/2; i++)
+	{
+		ROM[i] = ROM[i] ^ 0x3b90;
+
+		ROM[i] = bitswap<16>(ROM[i], 8, 7, 13,  15,  4,  5,  12,  10,
+									 3,  1,  11,  9,  6,  14,  0, 2);
+
+		ROM[i] = ((ROM[i] & 0xff00) >> 8) | ((ROM[i] & 0x00ff) << 8);
+	}
+
+	m_maincpu->set_alt_tile_addressing_hack(0);
+	m_maincpu->set_disallow_resolution_control();
+
+	// why do we need these? it will jump to 0 after the menu selection (prior to fadeout and bank select) otherwise, which can't be correct
+
+	ROM[0x19c9a / 2] = 0xF165;
+	ROM[0x19c9c / 2] = 0xF165;
+	ROM[0x19c9e / 2] = 0xF165;
+
+	ROM[0x19cb8 / 2] = 0xF165;
+	ROM[0x19cba / 2] = 0xF165;
+	ROM[0x19cbc / 2] = 0xF165;
+
+	ROM[0x19cd4 / 2] = 0xF165;
+	ROM[0x19cd6 / 2] = 0xF165;
+	ROM[0x19cd8 / 2] = 0xF165;
+
+
+}
```

Comment:
Please use lowercase hex literals consistently and clean up the trailing blank lines inside the function body.